Automattic/woocommerce-payments
# Display APM setup task on the WCPay Overview page

Fixes #2169

## Changes proposed in this Pull Request





Display the "Boost your sales by accepting new payment methods/Set up additional payment methods" task on the WCPay Overview page (`Payments > Overview`).

(The conditions for displaying the task described below are the same as on `WooCommerce > Home`.)

Either task is displayed only if the `_wcpay_feature_upe_additional_payment_methods` feature flag is enabled. Then:

* If `_wcpay_feature_upe_settings_preview` is enabled, the "Boost your sales" task will be displayed.
  * Marked as complete if either you go through the setup or `_wcpay_feature_upe` is enabled.
* If `_wcpay_feature_upe_settings_preview` is disabled and `_wcpay_feature_upe` is enabled, the "Set up additional methods" task will be displayed.
  * Marked as complete only if you go through the setup.

## Known issues

I was able to resolve an issue with the buttons not working (mentioned in the "Implementation details" section) and create a follow-up issue for removing the button thanks to information from @octaedro (p1625663464222300-slack-C9K5T0XSP). Thank you!

* The task list item design displays a button (which seems to be inconsistent with the desired design in the issue #2169).
  * This was implemented in https://github.com/woocommerce/woocommerce-admin/pull/7263 but doesn't yet seem to be available in WCPay.
  * Created follow-up issue: #2455
* To display the task, the code redirects to the task's WC > Home page. This has the consequence that clicking the "Back" button in the top-left corner of the page results in going to WC > Home even if you initiate the wizard on WCPay > Overview.
  * Created follow-up issue: #2454

This GIF demonstrates the issues plus it also demonstrates that the only clickable task component is the button: https://d.pr/i/0zS9C2

## Testing instructions

1. Clear the DB to make sure the task is not marked as completed:
  ```sql
  DELETE FROM `wp_options` WHERE `option_name` = 'wcpay_additional_methods_setup_completed'
  ```
2. Verify that:
  1. Tasks are displayed as described in the "Changes proposed in this Pull Request" section.
  2. Clicking the button on the WCPay > Overview page redirects to the task.
  3. Clicking the button on the WC > Home page was left intact.

## Screenshots

(The other task is just to illustrate the button always being displayed.)

### "Boost your sales" task (UPE preview enabled)

Pending task: 

![Markup on 2021-07-06 at 12:17:09](https://user-images.githubusercontent.com/1759681/124585587-b8dcda00-de55-11eb-976d-bd4e8b75e6a2.png)

Completed task:

![Markup on 2021-07-06 at 12:20:06](https://user-images.githubusercontent.com/1759681/124585601-bda18e00-de55-11eb-8dd4-8216d263d544.png)

### "Set up additional methods" task (UPE preview disabled, UPE enabled)

Pending task:

![Markup on 2021-07-06 at 12:23:58](https://user-images.githubusercontent.com/1759681/124585635-c72af600-de55-11eb-8612-83e8998a928f.png)

Completed task:

![Markup on 2021-07-06 at 12:23:12](https://user-images.githubusercontent.com/1759681/124585649-c98d5000-de55-11eb-975e-561e54186bdd.png)

## Implementation details

Since WCPay already has a mechanism for displaying a task list shared between `WC > Home` and `WCPay > Overview`, I have refactored the existing code used to create it to make use of this mechanism, moving the logic from `WC_Payments_Admin_Additional_Methods_Setup` to `WC_Payments_Admin` and removing the former class.

I have also refactored the existing task to use callbacks provided by a new context rather than setting `window.*` variables directly.

This PR also [contains a fix](https://github.com/Automattic/woocommerce-payments/pull/2358/files#diff-64da8dc3c200ed8ee47daef3203920aa6d1554e509a9eda184c959844b90991bR87) for a bug where, if meeting the conditions for the "Boost your sales" task to be displayed, the task was not marked as completed if UPE was enabled but you didn't go through the stepper. This was because we were checking for `'1' === isUpeEnabled` but the value returned from the back-end was boolean.

After a button on the WCPay > Overview task list was clicked (including for existing tasks), an error was thrown in the developer console and nothing else happened. In #2365, tasks' `onClick` props are mapped to `action` props supported by `woocommerce-admin`'s `<TaskList>`. I have [replicated this in this PR](https://github.com/Automattic/woocommerce-payments/pull/2358/files#diff-34ab3d28feff22ba3c825956998f112205530458a81a255ebfdfc293d8221269R46).





-------------------

  - Not added as this is a part of a larger feature.
